docs/features/source-generators.cookbook.md

```diff
@@ -0,0 +1,250 @@
+
+There are already well supported tools and techniques for acheiving these kinds of operations, and the source generators proposal is not aimed at replacing them.
+
+## Designs
+
+This section is broken down by user scenarios, with general solutions listed first, and more specific examples later on.
+
+### Generated class
+
+**User scenario:** As a generator author I want to be able to create a class that can be referenced from user code.
+
+**Solution:** Generate the class based on the data available in the compilation, and add the new source to compilation.
+
+**Example:**
+
+```csharp
+public class CustomGenerator : ISourceGenerator
+{
+    public void Execute(SourceGeneratorContext context)
+    {
+        context.AddSource("myGeneratedFile.cs", SourceText.From($@"
+namespace GeneratedNamespace
+{
+    public class GeneratedClass
+    {
+        public static void GeneratedMethod()
+        {
+            // generated code
+        }
+    }
+}"));
+    }
+}
+```
+
+```csharp
+public partial class UserClass
+{
+    public void UserMethod()
+    {
+        // call into a generated method
+        GeneratedNamespace.GeneratedClass.GeneratedMethod();
+    }
+}
+```
+
+### Single file transformation
+
+**User scenario:** As a generator author I want to be able to transform a single file into a C# representation.
+
+**Solution:** Use the additional files property of the `SourceGeneratorContext` to retreive the contents of the file, convert it to the C# representation and return it.
+
+**Example:**
+
+```csharp
+public class FileTransformGenerator : ISourceGenerator
+    {
+        public void Execute(SourceGeneratorContext context)
+        {
+            // find anything that matches our files
+            var myFiles = context.AnalyzerOptions.AdditionalFiles.Where(at => at.Path.EndsWith(".myextension"));
+            foreach (var file in myFiles)
+            {
+                var content = file.GetText(context.CancellationToken);
+
+                // do some transforms based on the file contenxt
+
+                string output = "namespace MyNS { public class MyClass { } }";
+                var sourceText = SourceText.From(output);
+
+                context.AddSource($"{file.Name}generated.cs", sourceText);
+            }
+        }
+    }
+```
+
+### Augment user code
+
+**User scenario:** As a generator author I want to be able to augment a users code with new functionality.
+
+**Solution:** Require the user to make the class you want to augment be a `partial class`, and mark it with e.g. a unique attribute, or name.
+Look for any classes marked for generation and generate the a matching `partial class` that contains the additional functionality.
+
+**Example:**
+
+```csharp
+public partial class UserClass
+{
+    public void UserMethod()
+    {
+        // call into a method inside the class
+        this.GeneratedMethod();
+    }
+}
+```
+
+```csharp
+public class AugmentingGenerator : ISourceGenerator
```

Comment:
nit: for user implemented code you'll want an abstract base type rather than an interface (because otherwise you can't really add new members without breaking them).